### PR TITLE
feat: pkam/onboarding design changes

### PR DESCRIPTION
**Pull request by murali-shris**
**- Reason for the changes**
- provide ability for callers of at_client to inject an instance of at chops which has implementation to sign pkam using private key from secure element.
- in at_client_service.dart , authenticate method also has onboarding logic which is confusing. Refactor code to separate onboarding and authentication logic.
**- What I did**
- changes for pkam using secure element and refactoring at_client_service
**- How I did it**
- added persistKeys method in at_client_spec.dart. This method will persist pkam and encryption keys to local secondary
- deprecated AtClientService since authenticate method was having onboarding logic and there is another method onboard. 
- deprecated class at_client_auth.dart. Moved the logic to new implementation at_client_service_impl.dart
- added an abstraction for  AtClientService which will have onboard, isOnboarded and authenticate methods. onboard will be called first time an atsign is activated from a client/app. authenticate will be called for subsequent login attempts
- added an abstraction AtAuthenticator for pkam and cram authentication
- implemented onboarding and authenticate methods in at_client_service_impl.dart
- added new class representations for AtOnboardingRequest, AtOnboardingResponse,AtAuthRequest, AtAuthResponse,AtKeysFileData

**- How to verify it**
- TODO. have to write unit and functional tests.

PENDING
* documentation
* some code changes related to authentication flow
* should sync be called explicitly after onboard

### CODE REVIEW

**Comment by gkc** [comment]
Hi @murali-shris - as this is draft, I'm guessing it's not ready for review. But if you are looking for a review, please can you add to the PR description giving some context as to why this was necessary, what options there were, why you chose this option, etc?

**Review by gkc — CHANGES_REQUESTED** [review]
More explanation in the PR description please

**Comment by murali-shris** [comment]
> More explanation in the PR description please

Sure Gary. I did a skeleton code to verify the design. 

Once the other PR is done - https://github.com/atsign-foundation/at_libraries/pull/301
I will do one more pass around the skeleton code, add the details and re-request review.

**Comment by murali-shris** [comment]
@gkc please review the design changes and new classes I have added.
I want to get an initial feedback before I start writing tests

File: packages/at_client_mobile/lib/src/at_client_service_v2.dart [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+class AtOnboardingResponse {
+  String atSign;
+  AtOnboardingResponse(this.atSign);
+  bool? isSuccessful;
+  AtKeysFileData? atKeysData;
+}
+
+enum PkamAuthMode { keysFile, sim }
```
Comment (gkc) [inline packages/at_client_mobile/lib/src/at_client_service_v2.dart]: Isn't this already defined in another package in at_libraries somewhere?

**Comment by gkc** [comment]
Hi @murali-shris 

The changes seem generally fine but I don't see a description of the 'why' - can you explain _why_ these changes are required please?

Also - is there any way to avoid the "AtClientServiceV2" name? I generally don't like giving versions in names. Is it impossible to evolve the current AtClientService?

File: packages/at_client_mobile/lib/src/at_client_service_v2.dart (hunk 1 above)
Comment (murali-shris) [reply packages/at_client_mobile/lib/src/at_client_service_v2.dart]: I have added PkamAuthMode to at_commons and removed the class from at_client_mobile package

**Comment by murali-shris** [comment]
> Hi @murali-shris
> 
> The changes seem generally fine but I don't see a description of the 'why' - can you explain _why_ these changes are required please?
> 
> Also - is there any way to avoid the "AtClientServiceV2" name? I generally don't like giving versions in names. Is it impossible to evolve the current AtClientService?

Current AtClientService will have references in mobile app code. It will be difficult to refactor the class without breaking changes.I have renamed AtClientServiceV2 to AtAuthService and AtClientServiceImpl to AtAuthServiceImpl.  Is this fine?

**Review by gkc — COMMENTED** [review]
LGTM

**Comment by gkc** [comment]
Hi @murali-shris what's next step for this PR?

**Comment by murali-shris** [comment]
Revisit this PR after APKAM changes

**Comment by murali-shris** [comment]
Closing this PR.Since we have refactored the code using at_auth package.

**Pull request closed without merging**